File: client/src/main/java/com/sumian/sd/buz/notification/NotificationListActivity.java
package com.sumian.sd.buz.notification;

import android.content.Context;
import android.content.Intent;
import android.view.View;

import com.blankj.utilcode.util.ActivityUtils;
import com.blankj.utilcode.util.LogUtils;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.sumian.common.base.BaseViewModelActivity;
import com.sumian.common.notification.NotificationUtil;
import com.sumian.common.utils.SettingsUtil;
import com.sumian.common.widget.CommonEmptyView;
import com.sumian.sd.R;
import com.sumian.sd.buz.notification.bean.Notification;
import com.sumian.sd.buz.stat.StatConstants;
import com.sumian.sd.widget.TitleBar;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;


public class NotificationListActivity extends BaseViewModelActivity<NotificationListPresenter>
        implements BaseQuickAdapter.OnItemClickListener, BaseQuickAdapter.RequestLoadMoreListener {

    public static final int REQUEST_CODE_OPEN_NOTIFICATION = 1;
    private TitleBar titleBar;
    private NotificationListAdapter mAdapter;
    private NotificationListHeadView mHeaderView;

    public static void launch(Context context) {
        ActivityUtils.startActivity(getLaunchIntent(context));
    }

    public static Intent getLaunchIntent(Context context) {
        return new Intent(context, NotificationListActivity.class);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_notification_list;
    }

    @NotNull
    @Override
    public String getPageName() {
        return StatConstants.page_notification_list;
    }

    @Override
    protected void initWidget() {
        super.initWidget();
        setMViewModel(new NotificationListPresenter(this));
        titleBar = findViewById(R.id.title_bar);
        titleBar.setOnBackClickListener(v -> finish());
        titleBar.setOnMenuClickListener(v -> markAllAsRead());
        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new NotificationListAdapter(this, null);
        recyclerView.setAdapter(mAdapter);
        mAdapter.setOnItemClickListener(this);
        mAdapter.setOnLoadMoreListener(this, recyclerView);
        mAdapter.setEmptyView(getEmptyView());
        initHeadView();
    }

    private void initHeadView() {
        boolean enabled = isNotificationEnabled();
        if (!enabled) {
            mHeaderView = new NotificationListHeadView(this);
            mHeaderView.setOnClickListener(v -> SettingsUtil.launchSettingActivityForResult(this, REQUEST_CODE_OPEN_NOTIFICATION));
            mAdapter.addHeaderView(mHeaderView);
        }
    }

    private boolean isNotificationEnabled() {
        return NotificationUtil.INSTANCE.areNotificationsEnabled(this);
    }

    private void removeHeadViewInNeeded() {
        boolean enabled = isNotificationEnabled();
        if (enabled && mHeaderView != null) {
            mAdapter.removeHeaderView(mHeaderView);
            mHeaderView = null;
        }
    }

    private View getEmptyView() {
        return new CommonEmptyView(this, null)
                .setImage(R.mipmap.ic_empty_state_alarm)
                .setTitle(R.string.notification_list_empty_title)
                .setMessage(R.string.notification_list_empty_desc);

    }


    @Override
    protected void initData() {
        super.initData();
        getMViewModel().loadData(true);
    }

    /**
     * 重复启动（点击notification）会进入该分支，需要刷新数据
     */
    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        mAdapter.setNewData(null);
        getMViewModel().loadData(true);
    }

    @Override
    public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
        Notification notification = mAdapter.getItem(position);
        if (notification == null) {
            return;
        }
        markAsRead(notification, position);
        launchActivityIfNeed(notification);
    }

    private void markAsRead(Notification notification, int position) {
        getMViewModel().readNotification(notification.getId(), notification.getDataId());
        notification.setReadAt((int) (System.currentTimeMillis() / 1000L));
        mAdapter.setData(position, notification);
    }

    private void markAllAsRead() {
        getMViewModel().readNotification("0", 0);
        List<Notification> data = mAdapter.getData();
        long currentTimeMillis = System.currentTimeMillis();
        for (Notification notification : data) {
            notification.setReadAt((int) (currentTimeMillis / 1000L));
        }
        mAdapter.notifyDataSetChanged();
    }

    private void launchActivityIfNeed(Notification notification) {
        String type = notification.getType();
        if (Notification.Companion.getTYPE_FOLLOW_UP_LIFE_NOTICE().equals(type)
                || Notification.Companion.getTYPE_FOLLOW_UP_REFERRAL_NOTICE().equals(type)) {
            return;
        }
        String scheme = notification.getData().getScheme();
        Intent intent = SchemeResolver.INSTANCE.schemeResolver(this, scheme);
        if (intent == null) {
            LogUtils.d("Unresolved scheme", scheme);
            return;
        }
        startActivity(intent);
    }

    public void onLoadMore(List<Notification> notificationList, boolean hasMore) {
        titleBar.setMenuVisibility(mAdapter.getItemCount() > 0 ? View.VISIBLE : View.GONE);
        if (notificationList == null) {
            notificationList = new ArrayList<>();
        }
        mAdapter.addData(notificationList);
        mAdapter.loadMoreComplete();
        mAdapter.setEnableLoadMore(hasMore);
    }

    public void onReadSuccess() {
    }

    @Override
    public void onLoadMoreRequested() {
        getMViewModel().loadData(false);
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE_OPEN_NOTIFICATION) {
            removeHeadViewInNeeded();
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    public void onBegin() {

    }

    public void onFinish() {

    }
}

File: client/src/main/java/com/sumian/sd/buz/notification/NotificationListActivity.kt
package com.sumian.sd.buz.notification

import android.content.Context
import android.content.Intent
import android.view.View
import com.blankj.utilcode.util.ActivityUtils
import com.blankj.utilcode.util.LogUtils
import com.sumian.common.base.BaseViewModel
import com.sumian.common.base.BaseViewModelActivity
import com.sumian.common.network.error.ErrorCode
import com.sumian.common.network.response.ErrorResponse
import com.sumian.module_core.async.AsyncCallback
import com.sumian.module_core.notification.Notification
import com.sumian.module_core.notification.NotificationListFragment
import com.sumian.module_core.notification.NotificationListResponse
import com.sumian.sd.R
import com.sumian.sd.app.AppManager
import com.sumian.sd.buz.patientdoctorim.ConversationListActivity
import com.sumian.sd.buz.stat.StatConstants
import com.sumian.sd.common.network.callback.BaseSdResponseCallback
import com.sumian.sd.common.utils.EventBusUtil
import kotlinx.android.synthetic.main.activity_notification_list.*

class NotificationListActivity : BaseViewModelActivity<BaseViewModel>(), NotificationListFragment.Host {

    private lateinit var mNotificationListFragment: NotificationListFragment

    override fun getLayoutId(): Int {
        return R.layout.activity_notification_list
    }

    override fun getPageName(): String {
        return StatConstants.page_notification_list
    }

    override fun initWidget() {
        super.initWidget()
        title_bar!!.setOnBackClickListener { v -> finish() }
        initFragment()
    }

    private fun initFragment() {
        val notificationListFragment = NotificationListFragment()
        supportFragmentManager.beginTransaction().replace(R.id.fragment_container, notificationListFragment).commit()
        mNotificationListFragment = notificationListFragment
    }

    override fun getNotificationList(page: Int, perPage: Int, callback: AsyncCallback<NotificationListResponse>) {
        val call = AppManager.getSdHttpService().getNotificationList(page, perPage)
        addCall(call)
        call.enqueue(object : BaseSdResponseCallback<NotificationListResponse>() {
            override fun onSuccess(response: NotificationListResponse?) {
                callback.onSuccess(response)
            }

            override fun onFailure(errorResponse: ErrorResponse) {
                callback.onFailed(errorResponse.code, errorResponse.message)
            }

            override fun onFinish() {
                callback.onFinish()
            }
        })
    }

    override fun readNotificationList(notificationId: String, dataId: Int, callback: AsyncCallback<Any>) {
        AppManager.getSdHttpService().readNotification(notificationId, dataId)
                .enqueue(object : BaseSdResponseCallback<Any>() {
                    override fun onSuccess(response: Any?) {
                        EventBusUtil.postStickyEvent(NotificationUnreadCountChangeEvent())
                        callback.onSuccess(response)
                    }

                    override fun onFailure(errorResponse: ErrorResponse) {
                        if (errorResponse.code == ErrorCode.STATUS_CODE_ERROR_UNKNOWN) {
                            callback.onSuccess(null) // 成功的response body 为空response code = 202, 医生版是204就不会，会走到onFail分支-_-||
                        } else {
                            callback.onFailed(errorResponse.code, errorResponse.message)
                        }
                    }
                })
    }

    override fun onNotificationClick(notification: com.sumian.module_core.notification.Notification) {
        launchActivityIfNeed(notification)
    }

    override fun showReadAll(show: Boolean) {
        title_bar!!.setMenuVisibility(if (show) View.VISIBLE else View.GONE)
    }

    override fun launchPatientDoctorMessageListActivity() {
        ConversationListActivity.launch()
    }

    override fun isDoctor(): Boolean {
        return false
    }

    /**
     * 重复启动（点击notification）会进入该分支，需要刷新数据
     */
    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        mNotificationListFragment.refreshData()
    }

    private fun launchActivityIfNeed(notification: Notification) {
        val type = notification.type
        if (Notification.TYPE_FOLLOW_UP_LIFE_NOTICE == type || Notification.TYPE_FOLLOW_UP_REFERRAL_NOTICE == type) {
            return
        }
        val scheme = notification.data.scheme
        val intent = SchemeResolver.schemeResolver(this, scheme)
        if (intent == null) {
            LogUtils.d("Unresolved scheme", scheme)
            return
        }
        startActivity(intent)
    }

    companion object {

        fun launch(context: Context) {
            ActivityUtils.startActivity(getLaunchIntent(context))
        }

        fun getLaunchIntent(context: Context): Intent {
            return Intent(context, NotificationListActivity::class.java)
        }
    }
}
